### ISSUE TITLE
Preview does not work on MS Win (neither WSL or Git Bash)

### ISSUE THREAD

**Issue by Grueslayer** [junegunn/fzf.vim#1212]
- [x] I have fzf 0.23.0 or above
- [x] I have read through https://github.com/junegunn/fzf.vim/blob/master/README.md
- [x] I have read through https://github.com/junegunn/fzf/blob/master/README-VIM.md
- [x] I have read through the manual page of fzf (`man fzf`)
- [x] I have searched through the existing issues

Preview will not run on MS Win systems (maybe Cygwin, couldn't test).

Let's have a look at the situation:

- VIM is running native on MSWin using path with backslash (and drive-letters and/or UNC path)
- FZF is running native on MSWin using same environment
- RG is running native on MSWin using same environment
- All others (like MRU lists) follow these directions

If now FZF calls the `preview.sh` script (by any bash.exe it founds) the argument is a *MS Win* path.

Let's assume we only found the WSL bash (where some exceptions are programmed into the fzf.vim code) it can only use files which are in the local directory. All other files contain the wrong path delimiter (or additional drive / UNC server names).

Only if we are on a bash (say the bash environment coming with the native GIT) that will work.

On WSL we need to *convert* the local MS Win path to a WSL path. Therefore exists a command called `wslpath` you can give a Windows path and receive a Linux path for the current WSL instance.

But there is a second special behaviour: VIM simulates the `~` known from un*x like systems as home directory for the current user. This is accomplished by checking against environment variables:

- `$HOME` (which is normally *not* set on a MS Win system
- `$HOMEDRIVE$HOMEPATH` (which default is the home of the current loggon user)

(It seems VIM does not check $HOMESHARE and does not use $PROFILEPATH)

So if you are using `History` you'll see p.e. `~\_vimrc`.

The WSL bash is a complete linux system which has its own `HOME` which has nothing in common with the windows users home (above).

So we need to tell the `preview.sh` what to do with that `~` in the path, too. The WSL has normally no access to MS Win environment vairiable but you can *forward* those. 

The *MS Win* environment variable (it's a list in un*x style) is to be set:

```
WSLENV=HOMEDRIVE/u:HOMEPATH/u:
```

Here we have two possible solutions for this:

1. The user sets those variables in the "System Enviromnent Variables" dialog from MS Win (WSL always has access to these variables)
2. If fzf.vim detects the WSL bash, it will check the environment  variable WSLENV for HOMEDRIVE and HOMEPATH content. if it does not exists, it adds those like above. The WSL bash has only access to it when called from VIM.

This a prove-of-concept running with WSL and Git bash:

```diff
diff --git a/bin/preview.sh b/bin/preview.sh
index 08a8c07..06587d2 100755
--- a/bin/preview.sh
+++ b/bin/preview.sh
@@ -22,6 +22,18 @@ if [[ -n "$CENTER" && ! "$CENTER" =~ ^[0-9] ]]; then
 fi
 CENTER=${CENTER/[^0-9]*/}
 
+#MS Win support
+if [ -z "$MSWINHOME" ]; then
+  MSWINHOME="$HOMEDRIVE$HOMEPATH"
+fi
+if grep -qEi "(Microsoft|WSL)" /proc/version &> /dev/null ; then
+  MSWINHOME="${MSWINHOME//\\/\\\\}"
+  FILE="${FILE/#\~\\/$MSWINHOME\\}"
+  FILE=`wslpath -u "$FILE"`
+elif [ ! -z "$MSWINHOME" ]; then
+  FILE="${FILE/#\~\\/$MSWINHOME\\}"
+fi
+
 FILE="${FILE/#\~\//$HOME/}"
 if [ ! -r "$FILE" ]; then
   echo "File not found ${FILE}"
```

**Comment by ndavidpsi** [comment]
I think I have the same issue. I have the following in my init.vim
```
command! -bang -nargs=? -complete=dir Files
      \ call fzf#vim#files(<q-args>, fzf#vim#with_preview({'options':
      \ ['--layout=default', '--info=inline']}), <bang>0)

```
If I then do :Files, I get this in my preview window (note that I have WSL's bash on PATH):
```
/bin/bash: At: No such file or directory
```

**Comment by ndavidpsi** [comment]
If I have Git's bash on PATH instead I get a similar message

**Comment by Grueslayer** [comment]
First of all fzf.vim only handles WSL if bash is found in .....\Windows\System32\bash.exe. Because this is one of the first directories in PATH you must set the path to the git bash BEFORE that entry to use it. You can check with `where bash` in a CMD prompt: first line is used. Any other shell not in windows system dir is treated as non WSL.

I think your problem is a different one, because the script is checking the file and gives you another error message. You can track it down by entering something like this `echo $LINENO: $FILE` in the preview.sh in multiple lines to see whet the FILE is and how far the progress.

**Comment by ndavidpsi** [comment]
I don't know which part refers to my case but when I said I was using the WSL2's bash I meant that the System32\bash.exe was the one on my path whereas for the git case I did what you said, put the Git's bash path before the system32 one. In the system32 one I get `/bin/bash: At: No such file or directory` when I do the `:Files` command, using the git one I get `/c/windows/system32/At: /c/windows/system32/At: cannot execute binary file` which is confusing...

**Comment by Grueslayer** [comment]
- Are you already using my fix?
- Did you entering the debug outputs like I wrote?
- Maybe we should discuss this at another place. I'm still thinking it's another issue.

**Comment by ndavidpsi** [comment]
No I haven't done that. I'll try it now and let you know.

**Comment by ndavidpsi** [comment]
@Grueslayer are you using cmd or pwsh as the Vim shell? I was using pwsh, I switched back to cmd and I got a different result. It now mentions some errors regarding the preview.sh file.

**Comment by ndavidpsi** [comment]
The thing is that I use pwsh because it accepts escaped spaces (or wtv it's called, I'm talking about the output of fzf) and cmd doesn't

**Comment by Grueslayer** [comment]
For Powershell you can find many other open tickets here, it's a different problem

**Comment by yanisdb** [comment]
Hi, 
I am having the same issue. Maybe we could make a preview.ps1 for windows? That way we wouldn't have to use any bash.exe and we wouldn't have any problem with paths?
Then we would just have to modify autoload/fzf/vim.vim to use that script instead of preview.sh.

If needed, I can translate the preview script for windows.

**Comment by Grueslayer** [comment]
Also the bash stuff above has another problem. If you type fast then some keystrokes will disappear (writing 'vimrc" often bocomes 'vmr' in my daily use).

Before I did the above fix (which is nearly a small fix, which can easily appoved by the maintainer but still isn't) I already made a small powershell. It has some less features

- only using bat.exe no other fallbacks
- everytime calls bat.exe (no exception for binary files, because 'file' or an something similar does not exist for Powershell on a standard system)

This is the fix

- it allows to set different scripts for preview and tags
- needed interpreter (like bash / powershell / perl) are mapped in a hash array

You'll need to tell fzf.vim to use the new preview.ps1 on your windows system like

```vim
if has('win32')
	let g:fzf_bin = { 'preview' : 'preview.ps1' }
endif
```

```diff
diff --git a/autoload/fzf/vim.vim b/autoload/fzf/vim.vim
index b56a2a6..f6d6324 100644
--- a/autoload/fzf/vim.vim
+++ b/autoload/fzf/vim.vim
@@ -32,9 +32,22 @@ let s:min_version = '0.23.0'
 let s:is_win = has('win32') || has('win64')
 let s:layout_keys = ['window', 'up', 'down', 'left', 'right']
 let s:bin_dir = expand('<sfile>:p:h:h:h').'/bin/'
+
 let s:bin = {
-\ 'preview': s:bin_dir.'preview.sh',
-\ 'tags':    s:bin_dir.'tags.pl' }
+\ 'preview': 'preview.sh',
+\ 'tags':    'tags.pl' }
+let s:bin = extend(s:bin, get(g:, 'fzf_bin', {}))
+for [key,val] in items(s:bin)
+    if val != '' && !(val =~ '[/\\]')
+      let s:bin[key] = (s:bin_dir).val
+    endif
+endfor
+
+let s:bin_ext = {
+\ 'sh'  : 'bash',
+\ 'pl'  : 'perl',
+\ 'ps1' : 'powershell'}
+"
 let s:TYPE = {'dict': type({}), 'funcref': type(function('call')), 'string': type(''), 'list': type([])}
 if s:is_win
   if has('nvim')
@@ -118,6 +131,7 @@ function! fzf#vim#with_preview(...)
   " Default spec
   let spec = {}
   let window = ''
+  let exefile = get(s:bin_ext, fnamemodify(s:bin.preview,':e'),'bash')
 
   let args = copy(a:000)
 
@@ -127,9 +141,9 @@ function! fzf#vim#with_preview(...)
     call remove(args, 0)
   endif
 
-  if !executable('bash')
+  if !executable(exefile)
     if !s:warned
-      call s:warn('Preview window not supported (bash not found in PATH)')
+      call s:warn('Preview window not supported ('.exefile.' not found in PATH)')
       let s:warned = 1
     endif
     return spec
@@ -152,10 +166,16 @@ function! fzf#vim#with_preview(...)
     let preview += ['--preview-window', window]
   endif
   if s:is_win
-    let is_wsl_bash = exepath('bash') =~? 'Windows[/\\]system32[/\\]bash.exe$'
-    let preview_cmd = 'bash '.(is_wsl_bash
-    \ ? substitute(substitute(s:bin.preview, '^\([A-Z]\):', '/mnt/\L\1', ''), '\', '/', 'g')
-    \ : escape(s:bin.preview, '\'))
+    let is_wsl_bash = exepath(exefile) =~? 'Windows[/\\]system32[/\\]bash.exe$'
+    if exefile == 'bash'
+      let preview_cmd = exefile.' '.(is_wsl_bash
+      \ ? substitute(substitute(s:bin.preview, '^\([A-Z]\):', '/mnt/\L\1', ''), '\', '/', 'g')
+      \ : escape(s:bin.preview, '\'))
+    elseif exefile == 'powershell'
+      let preview_cmd = exefile.' -f '.fzf#shellescape(s:bin.preview)
+    else
+      let preview_cmd = exefile.' '.fzf#shellescape(s:bin.preview)
+    endif
   else
     let preview_cmd = fzf#shellescape(s:bin.preview)
   endif
@@ -905,8 +925,9 @@ function! s:tags_sink(lines)
 endfunction
 
 function! fzf#vim#tags(query, ...)
-  if !executable('perl')
-    return s:warn('Tags command requires perl')
+  let exefile = get(s:bin_ext, fnamemodify(s:bin.tags,':e'),'perl')
+  if !executable(exefile)
+    return s:warn('Tags command requires '.exefile)
   endif
   if empty(tagfiles())
     call inputsave()
@@ -937,7 +958,7 @@ function! fzf#vim#tags(query, ...)
   let opts = v2_limit < 0 ? ['--algo=v1'] : []
 
   return s:fzf('tags', {
-  \ 'source':  'perl '.fzf#shellescape(s:bin.tags).' '.join(map(tagfiles, 'fzf#shellescape(fnamemodify(v:val, ":p"))')),
+  \ 'source':  exefile.' '.fzf#shellescape(s:bin.tags).' '.join(map(tagfiles, 'fzf#shellescape(fnamemodify(v:val, ":p"))')),
   \ 'sink*':   s:function('s:tags_sink'),
   \ 'options': extend(opts, ['--nth', '1..2', '-m', '--tiebreak=begin', '--prompt', 'Tags> ', '--query', a:query])}, a:000)
 endfunction
``` 

and this is the Powershell script

```ps1


if ($args.Count -lt 1)
{
  echo "usage: $PSCommandPath FILENAME[:LINENO][:IGNORED]"
  exit 1
}

if ($args[0] -match '^([A-Z]:)?([^:]*)(:([^:]*))?(:(.*))?$')
{
  $file = $Matches[1]+$Matches[2]
  $lineno = $Matches[4]
  $ignored = $Matches[6]
}
else
{
  $file = $args[0]
  $center = ''
  $ignored = ''
}

if ($lineno -and -not ($linenno -match '^[0-9]*$' ))
{
  echo "Argument LINENO='$center' is not a number"
  exit 20
}

$homedir = Get-Variable HOME -valueOnly 
if (-not $homedir)
{
  $homedir = (Get-Variable HOMEDRIVE -valueOnly) + (Get-Variable HOMEPATH -valueOnly)
}
if (-not ($homedir -match '\\$')) { $homedir = $homedir + '\' }


$file = ((($file -replace '\\\\', '\') -replace '\/', '\') -replace '~\\',$homedir)

if (-not (Test-Path $file -PathType Leaf))
{
  echo "File not found $file"
  exit 1
}

if (-not $lineno) { $lineno = '0' }

$batcmd = Get-Command "batcat" -ErrorAction SilentlyContinue
if ($batcmd -eq $null) { $batcmd = Get-Command "bat" -ErrorAction SilentlyContinue }
if ($batcmd -ne $null) { $bat = $batcmd.Name }

$batstyle = Get-Variable BAT_STYLE -valueOnly -ErrorAction SilentlyContinue
if (-not $batstyle) { $batstyle = 'numbers' }
$cmd = """$bat"" --style=""$batstyle"" --color=always --pager=never --highlight-line=$lineno ""$file"""

Invoke-Expression "& $cmd"

```

It is so sad, that we have no discussion with the maintainer. I would like to have 

a) my PR approved (that we have a correct running system)
b) my Powershell included (or at least an option to define your own script, like I started with my above patch) 

Best to you all who must live with Windows.

**Comment by jmwatte** [comment]
Not to hijack this, but is this related to the comment in closed issue

#https://github.com/junegunn/fzf.vim/issues/1210
The last remark.
Have the same problem.
 Minimal set up ...



```
set nocompatible
syntax enable
filetype plugin on
set hidden
let s:this_dir = expand('<sfile>:p:h')
call plug#begin(s:this_dir.'/plugged')
Plug 'junegunn/fzf', { 'do': { -> fzf#install() } }
Plug 'junegunn/fzf.vim'
call plug#end()

let &shell = has('win32') ? 'powershell' : 'pwsh'
set shellquote= shellpipe=\| shellxquote=
set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command
set shellredir=\|\ Out-File\ -Encoding\ UTF8
```

gives me errors in the preview window

``"WINDOS/system32/At:  cannot execute binary file``

remove the last 4 lines for the pwsh and no problems.

**Comment by Grueslayer** [comment]
Please, this is completely another problem.

1. Vim is calling the default shell (in your example Powershell).
2. FZF will be called in that environment
3. FZF will call the given Preview script (here bash.exe is prepended)
4. The Bash script is been called from FZF with given filename and runs bat.exe

I fixed 4. and partly 3. for running correctly with Windows (WSL and git shell).

You all have the problem, that in 1. the argument given to FZF is not correctly escaped because cmd.exe and PowerShell.exe handles argument different.

There are already PowerShell issues where it belongs to.

**Comment by jmwatte** [comment]
your welcome. Deleted.

**Comment by Acelya-9028** [comment]
Hello, I'm an another Windows user. Originally, I had two problems, but looking for a fix, I realized that the 2 overlap.

I use

```
Windows 10 Entreprise LTSC 10.0.17763 N/A version 17763
NVIM v0.5.0-dev+1022-g77a6049e0

set shell=pwsh shellpipe=\| shellxquote=
set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command
set shellredir=\|\ Out-File\ -Encoding\ UTF8
```

First problem:

The first window opening of fzf is slow like described [here](https://github.com/junegunn/fzf.vim/issues/864). Not all opening, only the first.
So I played a bit with `reltime()` and I found that `let s:bin.preview = split(system('for %A in ("'.s:bin.preview.'") do @echo %~sA'), "\n")[0]` take `0.379268` seconds. Glurps. Once commented, no more lag.

Second problem:

I have a bash executable without an installed linux subsystem because... So like @Grueslayer, I have to use an another wrapper for a working preview.

```
→ C:\Users\K2-18b› Get-Command bash

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Application     bash.exe                                           10.0.1776… C:\Windows\system32\bash.exe

→ C:\Users\K2-18b› bash
Le sous-système Windows pour Linux n'a pas de distributions installées.
Vous pouvez installer les distributions en visitant le Microsoft Store :
https://aka.ms/wslstore
```

Your fix @Grueslayer is great (thanks for that) but it's not perfect because it forces the used interpreter and if you don't want to use `powershell` but `pwsh`, or whatever, you come back to the starting point.
I think it's preferable to not force it so the user can choose his poison, like: `let g:fzf_bin = { 'preview' : 'pwsh -NoLogo -NoProfile -File stdpath('config').'/preview.ps1' }` or `let g:fzf_bin = { 'preview' : 'xXx_my_custom_preview_wrapper_xXx' }`. Same for tags, obviously.

The final result seems to be the same for you @Grueslayer, just, you have to replace `let g:fzf_bin = { 'preview' : 'preview.ps1' }` by `let g:fzf_bin = { 'preview' : 'powershell -f preview.ps1' }`.

Btw, while testing I realized another problem. As you can see in the screenshot below, the loop above assume `cmd.exe` as shell and don't work if Powershell is used.

![fzf_error](https://user-images.githubusercontent.com/45818681/106366201-8c6ae200-633a-11eb-8841-10983baebf48.png)

I think it's just better to force CMD for this specific case. I used `jobwait(jobstart())` and not `system()` because escaping is not the same depending `system()` uses CMD or Powershell... Moreover, this fix the slowness when `preview.sh` is used.

Diff:

```diff
diff --git a/autoload/fzf/vim.vim b/autoload/fzf/vim.vim
index 4d5952f..5e639ed 100644
--- a/autoload/fzf/vim.vim
+++ b/autoload/fzf/vim.vim
@@ -32,13 +32,16 @@ let s:min_version = '0.23.0'
 let s:is_win = has('win32') || has('win64')
 let s:layout_keys = ['window', 'up', 'down', 'left', 'right']
 let s:bin_dir = expand('<sfile>:p:h:h:h').'/bin/'
-let s:bin = {
+let s:bin = extend({
 \ 'preview': s:bin_dir.'preview.sh',
-\ 'tags':    s:bin_dir.'tags.pl' }
+\ 'tags':    s:bin_dir.'tags.pl' }, get(g:, 'fzf_bin', {}))
 let s:TYPE = {'dict': type({}), 'funcref': type(function('call')), 'string': type(''), 'list': type([])}
 if s:is_win
-  if has('nvim')
-    let s:bin.preview = split(system('for %A in ("'.s:bin.preview.'") do @echo %~sA'), "\n")[0]
+  if has('nvim') && s:bin.preview =~ '/preview.sh'
+    function! s:on_stdout(job_id, data, event) dict
+      let s:bin.preview = trim(a:data[0])
+    endfunction
+    call jobwait([jobstart(['cmd', '/c', '"for %A in ("'.s:bin.preview.'") do @echo %~sA"'], {'on_stdout' : function('s:on_stdout'), 'stdout_buffered' : v:true})])
   else
     let s:bin.preview = fnamemodify(s:bin.preview, ':8')
   endif
@@ -127,12 +130,14 @@ function! fzf#vim#with_preview(...)
     call remove(args, 0)
   endif
 
-  if !executable('bash')
-    if !s:warned
-      call s:warn('Preview window not supported (bash not found in PATH)')
-      let s:warned = 1
+  if s:bin.preview =~ '.*\.sh$'
+    if !executable('bash')
+      if !s:warned
+        call s:warn('Preview window not supported (bash not found in PATH)')
+        let s:warned = 1
+      endif
+      return spec
     endif
-    return spec
   endif
 
   " Placeholder expression (TODO/TBD: undocumented)
@@ -162,10 +167,14 @@ function! fzf#vim#with_preview(...)
     let preview += ['--preview-window', window]
   endif
   if s:is_win
-    let is_wsl_bash = exepath('bash') =~? 'Windows[/\\]system32[/\\]bash.exe$'
-    let preview_cmd = 'bash '.(is_wsl_bash
-    \ ? substitute(substitute(s:bin.preview, '^\([A-Z]\):', '/mnt/\L\1', ''), '\', '/', 'g')
-    \ : escape(s:bin.preview, '\'))
+    if s:bin.preview =~ '.*\.sh$'
+      let is_wsl_bash = exepath('bash') =~? 'Windows[/\\]system32[/\\]bash.exe$'
+      let preview_cmd = 'bash '.(is_wsl_bash
+      \ ? substitute(substitute(s:bin.preview, '^\([A-Z]\):', '/mnt/\L\1', ''), '\', '/', 'g')
+      \ : escape(s:bin.preview, '\'))
+    else
+      let preview_cmd = escape(s:bin.preview, '\')
+    endif
   else
     let preview_cmd = fzf#shellescape(s:bin.preview)
   endif
@@ -917,7 +926,7 @@ function! s:tags_sink(lines)
 endfunction
 
 function! fzf#vim#tags(query, ...)
-  if !executable('perl')
+  if !executable('perl') && s:bin.tags =~ '/tags.pl'
     return s:warn('Tags command requires perl')
   endif
   if empty(tagfiles())
@@ -949,7 +958,7 @@ function! fzf#vim#tags(query, ...)
   let opts = v2_limit < 0 ? ['--algo=v1'] : []
 
   return s:fzf('tags', {
-  \ 'source':  'perl '.fzf#shellescape(s:bin.tags).' '.join(map(tagfiles, 'fzf#shellescape(fnamemodify(v:val, ":p"))')),
+  \ 'source':  (s:bin.tags =~ '/tags.pl' ? 'perl' : s:bin.tags).' '.fzf#shellescape(s:bin.tags).' '.join(map(tagfiles, 'fzf#shellescape(fnamemodify(v:val, ":p"))')),
   \ 'sink*':   s:function('s:tags_sink'),
   \ 'options': extend(opts, ['--nth', '1..2', '-m', '-d', '\t', '--tiebreak=begin', '--prompt', 'Tags> ', '--query', a:query])}, a:000)
 endfunction
@@ -1088,7 +1097,7 @@ function! s:helptag_sink(line)
 endfunction
 
 function! fzf#vim#helptags(...)
-  if !executable('grep') || !executable('perl')
+  if !executable('grep') || (!executable('perl') && s:bin.tags =~ '/tags.pl')
     return s:warn('Helptags command requires grep and perl')
   endif
   let sorted = sort(split(globpath(&runtimepath, 'doc/tags', 1), '\n'))
@@ -1101,7 +1110,7 @@ function! fzf#vim#helptags(...)
   call writefile(['/('.(s:is_win ? '^[A-Z]:[\/\\].*?[^:]' : '.*?').'):(.*?)\t(.*?)\t/; printf(qq('.s:green('%-40s', 'Label').'\t%s\t%s\n), $2, $3, $1)'], s:helptags_script)
   return s:fzf('helptags', {
   \ 'source':  'grep -H ".*" '.join(map(tags, 'fzf#shellescape(v:val)')).
-    \ ' | perl -n '.fzf#shellescape(s:helptags_script).' | sort',
+    \ ' | '.(s:bin.tags =~ '/tags.pl' ? 'perl -n' : s:bin.tags).' '.fzf#shellescape(s:helptags_script).' | sort',
   \ 'sink':    s:function('s:helptag_sink'),
   \ 'options': ['--ansi', '+m', '--tiebreak=begin', '--with-nth', '..-2']}, a:000)
 endfunction
```

To recap, this fixes the slow first opening (with Powershell), allows the use of any command instead of Bash (or Powershell with the proposed fix) and fixes a broken path (with Powershell).

I can send a PR or I don't know, whatever if that can help.

**Comment by Grueslayer** [comment]
I'm really with you. Like I wrote, the PowerShell change was my first step I really haven't finished, just worked fine for me.

The Bash correction I made a PR for was only to keep the maintainers approach. Meanwhile the maintainer did some comments for my PR, I will include.

So after that we have a correct bash implementation.

I think (without having a closer look to your changes) this should become another PR.

If the maintainer don't want to maintain another preview script (which I could totally understand). We (or you) can make another package only containing the script and set the variable.

**Comment by blayz3r** [comment]
I am getting this weird message in the preview. 
![image](https://user-images.githubusercontent.com/12205650/108004991-73626200-6fc5-11eb-8686-0abad5220d87.png)

**Comment by djsavvy** [comment]
@Acelya-9028 thanks for the well-thought out fix. I'm running into this issue myself having set `pwsh` to be the shell in `nvim`. 

Is this PR ready to be opened? https://github.com/junegunn/fzf.vim/compare/master...Acelya-9028:master

**Comment by Acelya-9028** [comment]
@djsavvy I no longer use fzf.vim and I doubt it's merge one day but you can use the code and open an pr if you want :s
